keps/1224-lending-limit/README.md

```diff
@@ -0,0 +1,144 @@
+
+## Motivation
+
+Sometimes we want to keep some resources for guaranteed usage, so that when new jobs come into queue, they can be admitted immediately.
+
+Under the current implementation, we are using `BorrowingLimit` to define the maximum amount of quota that this ClusterQueue is allowed to borrow. But this may cause another ClusterQueue in the same cohort to run out of resources.
+
+Even if we set the `Preemption`, it still needs some time and spends a lot of unnecessary cost.
+
+So we need a reservation design for resource requests and security reasons: `LendingLimit`, to claim the quota allowed to lend, reserve a certain amount of resources to ensure that they will never be borrowed.
+
+### Goals
+
+- Implement `LendingLimit`, users can have a reservation of guaranteed resource by claiming the `LendingLimit`.
+
+### Non-Goals
+
+- Replace `BorrowingLimit` to some extent in the future.
+
+## Proposal
+
+In this proposal, `LendingLimit` is defined. The `ClusterQueue` will be limited to lend the specified quota to other ClusterQueues in the same cohort.
+
+### User Stories (Optional)
+
+#### Story 1
+
+In order to ensure the full utilization of resources, we generally set `BorrowingLimit` to max, but this may cause a `ClusterQueue` to run out of its all resources, and make the new incoming job slow to response for the slow preemption. This could be worse in a competitive cluster, jobs will borrow resources and be reclaimed over and over.
+
+So we want to reserve some resources for a `ClusterQueue`, so that any incoming jobs in the  `ClusterQueue` can get admitted immediately.
+
+### Risks and Mitigations
+
+With both BorrowingLimit and LendingLimit configured, one clusterQueue may not be able to borrow up to the limit just because we reserved the lending limit quota of resource. To aviod this, we can leave the BorrowingLimit unset to borrow as mush as we can.
+
+## Design Details
+
+### Kueue LendingLimit API
+
+Modify ResourceQuota API object:
+
+```go
+type ResourceQuota struct {
+	[...]
+
+	// lendingLimit is the maximum amount of quota for the [flavor, resource]
+	// combination that this ClusterQueue is allowed to lend the unused
+	// quota to other ClusterQueues in the same cohort.
+	// In total, at a given time, ClusterQueue can reserve
+	// a quantity of quota equals to nominalQuota - lendingLimit.
+	// If null, it means that there is no lending limit.
+	// If not null, it must be non-negative.
+	// lendingLimit must be null if spec.cohort is empty.
+	// +optional
+	LendingLimit *resource.Quantity `json:"lendingLimit,omitempty"`
+}
+```
+
+Also `Lended` is added to `ResourceUsage`.
+
+```go
+type ResourceUsage struct {
+	[...]
+
+    // Lended is a quantity of quota that is lended to other ClusterQueues in the cohort.
+    Lended resource.Quantity `json:"lended,omitempty"`
```

Comment:
I don't think we can safely implement this, because unused resources from multiple CQs compose a single pool of shareable resources.

So there is no concept of A is borrowing from B. A is borrowing from all the unused resource of B, C and any other CQs in the cohort.


I propose we drop this status field.

Reply:
We can drop this for now, but please add this as a Note for record.